<a href="https://colab.research.google.com/github/Usman-938/Assignments/blob/main/7_Tasks_week_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Read the content of the `/content/Tasks (1).docx` file and summarize the tasks and headings within it.

## Read Document Content

### Subtask:
Read the content of the `/content/Tasks (1).docx` file to access the tasks.


**Reasoning**:
To read a .docx file, I need to use the `python-docx` library. The first step is to import the `Document` class from the `docx` module and then use it to open the specified document.



In [1]:
from docx import Document

# Open the document
document = Document('/content/Tasks (1).docx')

print("Document loaded successfully.")

ModuleNotFoundError: No module named 'docx'

**Reasoning**:
The previous code failed because the `python-docx` library was not installed. I need to install it first using `!pip install python-docx`.



In [2]:
pip install python-docx
print("python-docx library installed successfully.")

SyntaxError: invalid syntax (ipython-input-565892798.py, line 1)

**Reasoning**:
The previous `pip install` command resulted in a `SyntaxError` because it was not prefixed with `!`. I need to add `!` to execute it as a shell command in the notebook.



In [3]:
!pip install python-docx
print("python-docx library installed successfully.")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 253.0/253.0 kB 6.0 MB/s eta 0:00:00
python-docx library installed successfully.


**Reasoning**:
Since the `python-docx` library has been successfully installed, I will now attempt to open the document again using the `Document` class, as was the original intention.



In [4]:
from docx import Document

# Open the document
document = Document('/content/Tasks (1).docx')

print("Document loaded successfully.")

Document loaded successfully.


**Reasoning**:
Now that the document is loaded, I need to extract its content, specifically the text from all paragraphs and headings, to fulfill the subtask requirement of reading the content.



In [5]:
full_text = []
for paragraph in document.paragraphs:
    full_text.append(paragraph.text)

# Also extract headings if any
headings = []
for paragraph in document.paragraphs:
    if paragraph.style.name.startswith('Heading'):
        headings.append(paragraph.text)

print("Extracted document text and headings.")

Extracted document text and headings.


## Extract Headings and Tasks

### Subtask:
Parse the document to identify all headings and the tasks described under each heading.


**Reasoning**:
To parse the document content, identify headings, and associate tasks with them, I will implement the provided logic to iterate through paragraphs and categorize text based on predefined heading keywords.



In [6]:
extracted_content = []
current_heading = None
current_tasks = []

heading_keywords = ['TASK', 'Real-World Problem:', 'Your Objective:', 'Tasks', 'Hint Code', 'What You Learn']

for paragraph in document.paragraphs:
    paragraph_text = paragraph.text.strip()

    is_heading = False
    for keyword in heading_keywords:
        if paragraph_text.startswith(keyword) and len(paragraph_text) > len(keyword) and (paragraph_text[len(keyword)] == ' ' or paragraph_text[len(keyword)] == '—' or paragraph_text[len(keyword)] == ':'):
            is_heading = True
            break
        elif paragraph_text == keyword: # For exact matches like 'Tasks' or 'What You Learn'
            is_heading = True
            break

    if is_heading:
        if current_heading is not None:
            extracted_content.append({
                'heading': current_heading,
                'content': current_tasks
            })
        current_heading = paragraph_text
        current_tasks = []
    elif paragraph_text:
        current_tasks.append(paragraph_text)

# Append the last heading and its content
if current_heading is not None:
    extracted_content.append({
        'heading': current_heading,
        'content': current_tasks
    })

print("Extracted content structure:")
for item in extracted_content:
    print(f"Heading: {item['heading']}")
    for task in item['content']:
        print(f"  - {task}")

Extracted content structure:
Heading: TASK 1 — SMART SURVEILLANCE (IMAGE DATA)
Heading: Real-World Problem: A shopping mall wants to detect crowd density using CCTV cameras.
Heading: Your Objective: Prepare image data correctly for an ML model.
Heading: Tasks
  - Simulate an RGB image (224, 224, 3)
  - Create a batch of 16 images
  - Print shapes and explain each dimension
Heading: Hint Code
  - import numpy as np
  - images = np.random.rand(16, 224, 224, 3)
  - print(images.shape)
Heading: What You Learn
  - Tensor shapes
  - Batch processing
  - Image ML pipeline
Heading: TASK 2 — VIDEO ACTIVITY RECOGNITION
Heading: Real-World Problem: A security system wants to detect running vs walking from videos.
Heading: Your Objective: Understand how video data is stored for ML.
Heading: Tasks
  - Simulate a 10-second video (30 FPS → 300 frames)
  - Store as a tensor
  - Add batch size
Heading: Hint Code
  - video = np.random.rand(1, 300, 224, 224, 3)
  - print(video.shape)
Heading: What You Le

## List Identified Tasks

### Subtask:
Present a clear list of all identified headings and their associated tasks, asking for further instructions on how to complete them.


## Identified Headings and Tasks

```python
output_markdown = ""
for item in extracted_content:
    output_markdown += f"### {item['heading']}\n"
    if item['content']:
        for task in item['content']:
            output_markdown += f"- {task}\n"
    else:
        output_markdown += "- No specific tasks listed under this heading.\n"

output_markdown += "\n\n---\n\nNow that the headings and tasks have been identified, please let me know which specific task you would like to proceed with, or if you have any other instructions.\n"

print(output_markdown)
```


**Reasoning**:
The subtask requires presenting the identified headings and tasks. The previous markdown block contained Python code to format this output. Now, I will execute this Python code to generate a clear, formatted list of all identified headings and their associated tasks, and then ask for further instructions.



In [7]:
output_markdown = ""
for item in extracted_content:
    output_markdown += f"### {item['heading']}\n"
    if item['content']:
        for task in item['content']:
            output_markdown += f"- {task}\n"
    else:
        output_markdown += "- No specific tasks listed under this heading.\n"

output_markdown += "\n\n---\n\nNow that the headings and tasks have been identified, please let me know which specific task you would like to proceed with, or if you have any other instructions.\n"

print(output_markdown)

### TASK 1 — SMART SURVEILLANCE (IMAGE DATA)
- No specific tasks listed under this heading.
### Real-World Problem: A shopping mall wants to detect crowd density using CCTV cameras.
- No specific tasks listed under this heading.
### Your Objective: Prepare image data correctly for an ML model.
- No specific tasks listed under this heading.
### Tasks
- Simulate an RGB image (224, 224, 3)
- Create a batch of 16 images
- Print shapes and explain each dimension
### Hint Code
- import numpy as np
- images = np.random.rand(16, 224, 224, 3)
- print(images.shape)
### What You Learn
- Tensor shapes
- Batch processing
- Image ML pipeline
### TASK 2 — VIDEO ACTIVITY RECOGNITION
- No specific tasks listed under this heading.
### Real-World Problem: A security system wants to detect running vs walking from videos.
- No specific tasks listed under this heading.
### Your Objective: Understand how video data is stored for ML.
- No specific tasks listed under this heading.
### Tasks
- Simulate a 10-sec

## Final Task

### Subtask:
Conclude the task by summarizing the identified tasks and preparing for the next steps based on user feedback.


## Summary:

### Data Analysis Key Findings

*   The `python-docx` library was successfully installed after an initial `SyntaxError` due to a missing `!` prefix for the `pip install` command.
*   The `/content/Tasks (1).docx` file was successfully loaded and its content extracted.
*   The document parsing mechanism successfully identified 7 main tasks, along with their associated real-world problems, objectives, specific tasks, hint code, and learning outcomes, using predefined keywords.
*   The identified tasks and their sub-components were structured and presented in a clear markdown format for easy review.

### Insights or Next Steps

*   The system is now ready to proceed with any of the identified tasks from the `/content/Tasks (1).docx` document based on user selection.
*   This approach demonstrates an effective method for automating the extraction and structuring of task-related information from document files.


# Task
Simulate a batch of 16 RGB images (224x224 pixels each) using NumPy, then print and explain the shape of the resulting tensor, detailing what each dimension represents for image processing.

## TASK 1: Simulate Image Data

### Subtask:
Simulate a batch of 16 RGB images (224x224 pixels each) using NumPy, then print and explain the shape of the resulting tensor, detailing what each dimension represents for image processing.


**Reasoning**:
First, I need to import the `numpy` library and create the simulated batch of images as a NumPy array with the specified dimensions. Then, I will print its shape.



In [8]:
import numpy as np

# Create a NumPy array named `images` that simulates a batch of 16 RGB images, each with dimensions 224x224 pixels.
images = np.random.rand(16, 224, 224, 3)

# Print the shape of the `images` tensor
print("Shape of the images tensor:", images.shape)

Shape of the images tensor: (16, 224, 224, 3)


```markdown
The shape of the `images` tensor is `(16, 224, 224, 3)`.

Here's what each dimension represents:

*   **First dimension (16):** This is the **batch size**, indicating that there are 16 individual images in this collection.
*   **Second dimension (224):** This represents the **height** of each image in pixels.
*   **Third dimension (224):** This represents the **width** of each image in pixels.
*   **Fourth dimension (3):** This represents the **color channels** of each image. For an RGB image, there are 3 channels (Red, Green, Blue).
```

The shape of the `images` tensor is `(16, 224, 224, 3)`.

Here's what each dimension represents:

*   **First dimension (16):** This is the **batch size**, indicating that there are 16 individual images in this collection.
*   **Second dimension (224):** This represents the **height** of each image in pixels.
*   **Third dimension (224):** This represents the **width** of each image in pixels.
*   **Fourth dimension (3):** This represents the **color channels** of each image. For an RGB image, there are 3 channels (Red, Green, Blue).


## TASK 2: Simulate Video Data

### Subtask:
Simulate a video tensor representing a 10-second video at 30 FPS (300 frames), with each frame being an RGB image of 224x224 pixels, including a batch size of 1. Print and explain the shape of the video tensor, focusing on the time dimension.


**Reasoning**:
To simulate the video tensor and print its shape, I will use NumPy to create an array with the specified dimensions: batch size (1), frames (10 seconds * 30 FPS = 300), height (224), width (224), and color channels (3 for RGB).



In [9]:
import numpy as np

# Simulate a video tensor representing a 10-second video at 30 FPS (300 frames),
# with each frame being an RGB image of 224x224 pixels, and a batch size of 1.
video_data = np.random.rand(1, 300, 224, 224, 3)

# Print the shape of the video_data tensor
print("Shape of the video tensor:", video_data.shape)

Shape of the video tensor: (1, 300, 224, 224, 3)


The shape of the `video_data` tensor is `(1, 300, 224, 224, 3)`.

Here's what each dimension represents:

*   **First dimension (1):** This is the **batch size**, indicating that there is 1 video in this batch.
*   **Second dimension (300):** This is the **time dimension** or **number of frames**, representing 300 frames (10 seconds * 30 FPS) for the video.
*   **Third dimension (224):** This represents the **height** of each video frame in pixels.
*   **Fourth dimension (224):** This represents the **width** of each video frame in pixels.
*   **Fifth dimension (3):** This represents the **color channels** of each video frame. For an RGB image/frame, there are 3 channels (Red, Green, Blue).

## TASK 3: Simulate Churn Probability and Thresholding

### Subtask:
Generate 20 random customer churn probability predictions. Apply a decision threshold (e.g., 0.7) to identify and count customers classified as high-risk for churn.


**Reasoning**:
To simulate customer churn predictions and identify high-risk customers, I will use NumPy to generate random probabilities, define a threshold, and then filter the predictions to count those exceeding the threshold.



In [10]:
import numpy as np

# Generate 20 random customer churn probability predictions
churn_predictions = np.random.rand(20)

# Define a decision threshold
threshold = 0.7

# Identify customers with churn probability greater than the threshold
high_risk_customers = churn_predictions[churn_predictions > threshold]

# Print the total number of customers classified as high-risk for churn
print(f"Original churn predictions:\n{churn_predictions}\n")
print(f"High-risk churn predictions (above {threshold}):\n{high_risk_customers}\n")
print(f"Number of high-risk customers: {len(high_risk_customers)}")

Original churn predictions:
[0.59234904 0.01398458 0.70393108 0.58159878 0.18871885 0.64742848
 0.03265852 0.84057551 0.20301032 0.46104498 0.27941152 0.61422759
 0.02867997 0.05024636 0.80603873 0.49006242 0.06597528 0.73397109
 0.83517175 0.82330639]

High-risk churn predictions (above 0.7):
[0.70393108 0.84057551 0.80603873 0.73397109 0.83517175 0.82330639]

Number of high-risk customers: 6


## TASK 4: Calculate Feature Correlation

### Subtask:
Create two numerical lists representing advertising budget and sales data. Calculate the Pearson correlation coefficient between these two features using NumPy and interpret the result in terms of feature relevance.


**Reasoning**:
To address the subtask, I will first create the specified numerical lists for advertising budget and sales, convert them into NumPy arrays, and then compute their Pearson correlation coefficient using `np.corrcoef()` as instructed. Finally, I will print the resulting correlation matrix.



In [11]:
import numpy as np

# Create two numerical lists representing advertising budget and sales data
ads = [100, 200, 300, 400, 500]
sales = [10, 20, 30, 40, 50]

# Convert lists to NumPy arrays
ads_array = np.array(ads)
sales_array = np.array(sales)

# Calculate the Pearson correlation coefficient
correlation_matrix = np.corrcoef(ads_array, sales_array)

# Print the correlation matrix
print("Correlation Matrix:")
print(correlation_matrix)

Correlation Matrix:
[[1. 1.]
 [1. 1.]]


The correlation matrix `[[1. 1.] [1. 1.]]` shows the Pearson correlation coefficients between the advertising budget and sales data.

Specifically, the off-diagonal elements (which are `1.`) represent the correlation between `ads` and `sales`.
A correlation coefficient of `1.0` indicates a **perfect positive linear relationship** between the two variables.

**Interpretation in terms of feature relevance:**

This result suggests that there is a very strong and direct relationship between the advertising budget and sales. As the advertising budget increases, sales also increase proportionally. In a real-world scenario, this would mean that the advertising budget is a highly relevant feature for predicting or explaining sales.

## TASK 5: Simulate Model Learning

### Subtask:
Simulate 10 steps of a simple gradient descent algorithm to minimize a quadratic loss function, starting with an initial weight and learning rate. Print the weight and loss value at each step to observe the learning process.


**Reasoning**:
To simulate the gradient descent algorithm, I will create a code block that initializes the weight and learning rate, then iterates 10 times to update the weight and calculate the loss at each step, printing the results as instructed.



In [12]:
import numpy as np

# Initialize a weight and learning rate
w = 20
lr = 0.1

print("Simulating 10 steps of gradient descent:")
print(f"Initial Weight: {w:.2f}")

# Loop 10 times to simulate 10 steps of gradient descent
for i in range(1, 11):
    # Calculate the current loss (quadratic function w^2)
    loss = w**2

    # Calculate the gradient of the loss with respect to the weight (d(w^2)/dw = 2w)
    gradient = 2 * w

    # Update the weight using the gradient descent formula
    w = w - lr * gradient

    # Print the step, updated weight, and loss
    print(f"Step {i}: Weight = {w:.2f}, Loss = {loss:.2f}")

Simulating 10 steps of gradient descent:
Initial Weight: 20.00
Step 1: Weight = 16.00, Loss = 400.00
Step 2: Weight = 12.80, Loss = 256.00
Step 3: Weight = 10.24, Loss = 163.84
Step 4: Weight = 8.19, Loss = 104.86
Step 5: Weight = 6.55, Loss = 67.11
Step 6: Weight = 5.24, Loss = 42.95
Step 7: Weight = 4.19, Loss = 27.49
Step 8: Weight = 3.36, Loss = 17.59
Step 9: Weight = 2.68, Loss = 11.26
Step 10: Weight = 2.15, Loss = 7.21


## TASK 6: Compare Image vs. Video Memory Cost

### Subtask:
Create a single RGB image tensor (224x224 pixels) and a video tensor (30 frames, each 224x224 RGB). Calculate and compare the memory usage in bytes for both tensors using NumPy's `.nbytes` attribute.


**Reasoning**:
To complete the subtask, I will create both an image and a video tensor using NumPy, calculate their memory usage in bytes, and then print these values along with a comparison statement.



In [13]:
import numpy as np

# Create a single RGB image tensor (224x224 pixels)
image_tensor = np.random.rand(224, 224, 3)

# Create a video tensor (30 frames, each 224x224 RGB)
video_tensor = np.random.rand(30, 224, 224, 3)

# Calculate memory usage in bytes for both tensors
image_memory_bytes = image_tensor.nbytes
video_memory_bytes = video_tensor.nbytes

# Print memory usages and comparison
print(f"Memory usage for image tensor: {image_memory_bytes} bytes")
print(f"Memory usage for video tensor: {video_memory_bytes} bytes")

print("\nComparison:")
if video_memory_bytes > image_memory_bytes:
    ratio = video_memory_bytes / image_memory_bytes
    print(f"The video tensor uses {ratio:.2f} times more memory than the image tensor.")
elif image_memory_bytes > video_memory_bytes:
    ratio = image_memory_bytes / video_memory_bytes
    print(f"The image tensor uses {ratio:.2f} times more memory than the video tensor.")
else:
    print("Both tensors use the same amount of memory.")

Memory usage for image tensor: 1204224 bytes
Memory usage for video tensor: 36126720 bytes

Comparison:
The video tensor uses 30.00 times more memory than the image tensor.


The comparison shows that the **video tensor uses 30.00 times more memory than the image tensor**.

This significant difference is directly attributable to the **time dimension** (or number of frames) in the video tensor. While a single image is a 3D array (height, width, channels), a video is essentially a sequence of images over time, making it a 4D array (frames, height, width, channels) or 5D if a batch dimension is included. In this specific simulation:

*   **Image Tensor:** `(224, 224, 3)` represents one frame. Its memory usage is `224 * 224 * 3` units (assuming each value takes 1 unit of memory).
*   **Video Tensor:** `(30, 224, 224, 3)` represents 30 frames. Its memory usage is `30 * 224 * 224 * 3` units.

Therefore, the video tensor's memory requirement is proportional to the number of frames it contains relative to a single image, which in this case is 30 times greater. This illustrates why processing video data is generally more computationally and memory intensive than processing individual images, especially in deep learning applications.

## TASK 7: Evaluate Business Decision Thresholds

### Subtask:
Generate 50 random fraud probability scores. Test different decision thresholds (0.5, 0.7, 0.9) to count how many transactions would be flagged as fraudulent at each threshold, demonstrating the trade-offs in business decisions.


**Reasoning**:
To generate random fraud probability scores and apply different thresholds as requested in the subtask, I will use a code block to import NumPy, generate the scores, iterate through the thresholds, count the flagged cases, and print the results.



In [14]:
import numpy as np

# Generate 50 random fraud probability scores
fraud_scores = np.random.rand(50)

# Define a list of thresholds to evaluate
thresholds = [0.5, 0.7, 0.9]

print("Evaluating fraud detection with different thresholds:")
for threshold in thresholds:
    # Identify fraud scores greater than the current threshold
    flagged_cases = fraud_scores[fraud_scores > threshold]

    # Count the number of flagged cases
    num_flagged = len(flagged_cases)

    # Print the threshold and the number of flagged cases
    print(f"  Threshold {threshold}: {num_flagged} transactions flagged as fraudulent.")

print("\nAll fraud scores:")
print(fraud_scores)

Evaluating fraud detection with different thresholds:
  Threshold 0.5: 22 transactions flagged as fraudulent.
  Threshold 0.7: 11 transactions flagged as fraudulent.
  Threshold 0.9: 3 transactions flagged as fraudulent.

All fraud scores:
[4.94447646e-02 5.23574086e-01 5.18865670e-01 5.69157912e-01
 1.96959615e-01 6.67124575e-01 6.80171926e-01 7.33106918e-01
 2.60168599e-01 8.73460225e-01 4.80288147e-01 4.51138378e-01
 8.94522933e-01 4.20804375e-01 8.66458562e-01 3.48530540e-01
 1.60003385e-01 8.11381852e-03 9.14594570e-01 9.23971487e-01
 3.17829665e-01 4.37382520e-01 7.02560972e-01 6.76496101e-02
 5.74877681e-01 3.57706247e-01 4.34605747e-01 7.53795811e-01
 7.53475459e-02 1.25048148e-01 8.88996574e-01 3.27714008e-01
 2.00810555e-01 4.80002187e-01 4.11974928e-01 6.82882008e-01
 1.73433138e-02 2.35448429e-01 4.65709852e-01 5.91785521e-01
 8.75039439e-01 8.87276225e-02 8.14489162e-04 5.75818344e-01
 9.32681110e-01 4.24841725e-02 3.14483713e-01 6.86584925e-01
 2.22139700e-01 6.26946931e-

The output demonstrates the impact of choosing different thresholds for identifying fraudulent transactions:

*   **Threshold 0.5**: With a threshold of 0.5, **22 transactions** were flagged as fraudulent. This is the most lenient threshold, meaning it catches more potential fraud. However, it also likely includes more false positives (legitimate transactions incorrectly flagged as fraud), leading to higher operational costs (e.g., investigating more cases) and potentially inconveniencing more customers.

*   **Threshold 0.7**: Raising the threshold to 0.7 resulted in **11 transactions** being flagged. This is a more conservative approach. It reduces the number of flagged cases, which can decrease operational costs and false positives. However, it also means that some genuinely fraudulent transactions with scores between 0.5 and 0.7 might be missed (false negatives).

*   **Threshold 0.9**: The most stringent threshold of 0.9 flagged only **3 transactions**. This approach minimizes false positives and associated costs, but it comes at the risk of missing a significant number of fraudulent transactions (high false negatives). Only the most obvious or high-confidence fraud cases are identified.

**Trade-offs in Business Decisions:**

This simulation highlights the critical trade-off in business decisions based on predictive models:

*   **Lower Thresholds (e.g., 0.5)**: Increase recall (catch more fraud) but decrease precision (more false positives, higher operational cost, customer inconvenience).
*   **Higher Thresholds (e.g., 0.9)**: Increase precision (fewer false positives, lower operational cost) but decrease recall (miss more fraud, higher financial loss due to undetected fraud).

The optimal threshold depends on the specific business context, the cost of false positives versus false negatives, and the acceptable level of risk.

## Final Task

### Subtask:
Summarize the completion of all 7 tasks and provide an overview of the outcomes and insights gained from each.


## Summary:

### Data Analysis Key Findings

*   **Task 1: Image Data Simulation**
    A batch of 16 RGB images (224x224 pixels each) was successfully simulated using NumPy, resulting in a tensor of shape `(16, 224, 224, 3)`. Each dimension clearly represents the batch size, height, width, and RGB color channels, respectively.
*   **Task 2: Video Data Simulation**
    A video tensor, representing a 10-second video at 30 FPS (300 frames) with each frame as an RGB image of 224x224 pixels and a batch size of 1, was simulated. The resulting tensor had a shape of `(1, 300, 224, 224, 3)`, explicitly demonstrating the addition of a time/frame dimension.
*   **Task 3: Churn Probability Thresholding**
    Out of 20 random customer churn probability predictions, 6 customers were identified as high-risk for churn when a decision threshold of 0.7 was applied.
*   **Task 4: Feature Correlation**
    For the provided advertising budget and sales data, the Pearson correlation coefficient was calculated as `1.0`, indicating a perfect positive linear relationship between the two variables.
*   **Task 5: Model Learning Simulation**
    A 10-step gradient descent simulation for a quadratic loss function, starting with an initial weight of 20 and a learning rate of 0.1, showed the weight decreasing from 20.00 to 2.15 and the loss reducing consistently from 400.00 to 7.21.
*   **Task 6: Memory Cost Comparison (Image vs. Video)**
    A single RGB image tensor (224x224 pixels) consumed 1,204,224 bytes of memory, while a video tensor (30 frames, each 224x224 RGB) consumed 36,126,720 bytes. The video tensor used 30.00 times more memory than the image tensor, directly proportional to its number of frames.
*   **Task 7: Business Decision Thresholds**
    Evaluating 50 random fraud probability scores against different thresholds showed: 22 transactions flagged at threshold 0.5, 11 at 0.7, and 3 at 0.9. This clearly illustrated the trade-offs between catching more cases (lower threshold) and reducing false positives (higher threshold).

### Insights or Next Steps

*   Understanding data dimensionality is crucial for efficient data representation and memory management, especially when transitioning from static images to dynamic video sequences.
*   Decision thresholds and model learning parameters like learning rates directly impact business outcomes and model performance; careful tuning is essential to balance precision, recall, and operational costs.
